### ISSUE TITLE
[Enhancement] Receiving push notification while Android app is in foreground mode

### ISSUE THREAD

**Issue by eyeveye** [thomasgalliker/Plugin.FirebasePushNotifications#57]
#### Summary
When receiving push notification while the App is opening, it will not have an notification in the system bar. Currently need to manually create an extra popup to show the notification when the App is at foreground.

#### Intended Use Case
When app is at foreground, and push notification received it should show in the system bar. Would be good to have the notification show in the system bar.

**Comment by sgreifeneder** [comment]
+1, I was also searching for this!

**Comment by thomasgalliker** [comment]
This is already available (for Android). I‘m not sure if such a thing also works on iOS? Does anyone know?

**Comment by sgreifeneder** [comment]
Only if we override the NotificationBuilder for Android and change the logic for ShouldHandleNotificationReceived. 
Maybe a flag would be nice to handle this.

**Comment by sovdchains** [comment]
This currently is possible by setting `priority` to `high` in push notification data

**Comment by sgreifeneder** [comment]
Only works on iOS

**Comment by sovdchains** [comment]
Oh you're right, I haven't been aware of this.

**Comment by thomasgalliker** [comment]
Strange. Can you show me the notification message which you send?
As @sovdchains said, you can set the priority flag to high which SHOULD create a local notification message - even when the app is in foreground mode. If this is not the case, it's clearly a bug.

Does such a notification message work?
```
{
    "message": {
        "token": "{{fcm_token}}",
        "notification": {
            "title": "Notification title",
            "body": "Notification body"
        },
        "data": {
            "priority": "high"
        }
    }
}
```

(Btw. you can find a decent collection of postman http calls with several different notification messages [here](https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/blob/develop/Docs/FCM%20Plugin.FirebasePushNotifications.postman_collection.json)).

**Comment by sovdchains** [comment]
@thomasgalliker after @sgreifeneder told me it does not work on Android I tried to recreate the issue and have been successful with being unsuccessful to show the notification on Android while the app is in foreground. The `priority` flag really seems to only work on iOS. I also can't find any corresponding code for Android like [on iOS](https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/blob/bab92d182e3b05795cd00d6e993b961fd1617a8a/Plugin.FirebasePushNotifications/Platforms/iOS/FirebasePushNotificationManager.cs#L251)

**Comment by thomasgalliker** [comment]
Ok. Did you test with the message I posted above? I'm curious what message you used.

The code that should care about priority=high on Android can be found [here](https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/blob/bab92d182e3b05795cd00d6e993b961fd1617a8a/Plugin.FirebasePushNotifications/Platforms/Android/NotificationBuilder.cs#L54). So, the bug must be in this area too.

**Comment by sgreifeneder** [comment]
But 2 lines [above](https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/blob/bab92d182e3b05795cd00d6e993b961fd1617a8a/Plugin.FirebasePushNotifications/Platforms/Android/NotificationBuilder.cs#L52) it explicitly checks for `if (isInForeground == false)`, or?

**Comment by thomasgalliker** [comment]
That's wrong. I'll fix it.

**Comment by sovdchains** [comment]
> Ok. Did you test with the message I posted above? I'm curious what message you used.
> 
> The code that should care about priority=high on Android can be found [here](https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/blob/bab92d182e3b05795cd00d6e993b961fd1617a8a/Plugin.FirebasePushNotifications/Platforms/Android/NotificationBuilder.cs#L54). So, the bug must be in this area too.

I did not test your exact payload, but I checked that the data is set to the same value. I'll wait for your mentioned fix and see if this helps out. If not I'll try your raw payload.

**Comment by thomasgalliker** [comment]
Thanks for the reminder. I removed the IsInForeground check in NotificationBuilder but that does not yet seem to solve the problem.

**Comment by thomasgalliker** [comment]
Just to summarize how I managed to send a high-priority notification which is also displayed if the Android app runs in foreground mode:

1. You need to create a notification channel which has Importance=High (or above). If your notification channel has lower priority you still receive the notification message, but it will appear in the notification tray - and not as a foreground notification. On top of that, a warning is logged.
```csharp
new NotificationChannelRequest
{
	ChannelId = "default_channel_id",
	ChannelName = "Default Channel",
	Description = "The default notification channel",
	LockscreenVisibility = NotificationVisibility.Public,
	Importance = NotificationImportance.High,
	IsDefault = true,
};
```
If you have no notification channels (not even the default notification channel) you cannot use NotificationBuilder. An error is logged that informs you about this situation.

2. Run the app in foreground mode. Make sure the notification permission is granted.
3. Send an FCM notification message with data-flag "priority: high":
```json
{
    "message": {
        "token": "{{fcm_token}}",
        "notification": {
            "title": "Notification title",
            "body": "Notification body"
        },
        "data": {
            "priority": "high"
        }
    }
}
```

**Comment by thomasgalliker** [comment]
@eyeveye @sgreifeneder @sgreifeneder thank you for raising this issue and for the patience with me. I hope you understand that this is by far not the only construction site I'm working on.

If any of you could try the [latest 2.4.x pre-release](https://www.nuget.org/packages/Plugin.FirebasePushNotifications/2.4.17-pre) and let me know if the problem was solved, I'd appreciate it.
If you find more suspicious code please do not hesitate to create a new issue.

**Comment by sovdchains** [comment]
@thomasgalliker Thank you for providing us with those awesome libraries of yours and maintaining them so well! I will try this on monday!

**Comment by sovdchains** [comment]
@thomasgalliker I just validated that the issue is gone on a Samsung Tab3 running Android 8.1. Thanks again for the fix!

**Comment by vchelaru** [comment]
Does the fix in the latest 2.4.x pre-release also require the creation of a new NotificationChannelRequest? If so, where does the creation of NotificationChannelRequest go? I'm new to push notifications so it's not clear if this goes in my maui app or if this is done on the backend where the push notification request is sent to firebase.
